atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -229,42 +301,44 @@ def get_cluster_info(self, cluster, platform):
 
         conf = Configuration(**kwargs)
         osbs = OSBS(conf, conf)
-        try:
-            current_builds = self.get_current_builds(osbs)
-        except OsbsException as e:
-            # If the build is canceled reraise the error
-            if isinstance(e.cause, BuildCanceledException):
-                raise e
 
-            self.log.exception("Error occurred while listing builds on %s",
-                               cluster.name)
-            return ClusterInfo(cluster, platform, osbs, self.UNREACHABLE_CLUSTER_LOAD)
+        current_builds = self.get_current_builds(osbs)
 
         load = current_builds / cluster.max_concurrent_builds
         self.log.debug('enabled cluster %s for platform %s has load %s and active builds %s/%s',
                        cluster.name, platform, load, current_builds, cluster.max_concurrent_builds)
         return ClusterInfo(cluster, platform, osbs, load)
 
-    def choose_cluster(self, platform):
-        config = get_config(self.workflow)
-        clusters = [self.get_cluster_info(cluster, platform) for cluster in
-                    config.get_enabled_clusters_for_platform(platform)]
-
-        if not clusters:
-            raise RuntimeError('No clusters found for platform {}!'
-                               .format(platform))
-
-        reachable_clusters = [cluster for cluster in clusters
-                              if cluster.load != self.UNREACHABLE_CLUSTER_LOAD]
-
-        if not reachable_clusters:
-            raise RuntimeError('All clusters for platform {} are unreachable!'
-                               .format(platform))
+    def get_clusters(self, platform, retry_contexts, all_clusters):
+        ''' return clusters sorted by load. '''
+
+        possible_cluster_info = {}
+        possible_clusters = set()
+        candidates = set(copy.copy(all_clusters))
+        while candidates and not possible_clusters:
+            for cluster in sorted(candidates, key=attrgetter('priority')):
+                ctx = retry_contexts[cluster.name]
+                if ctx.in_retry_wait:
+                    continue
+                try:
+                    cluster_info = self.get_cluster_info(cluster, platform)
+                    possible_cluster_info[cluster.name] = cluster_info
+                    # TODO: this is a problem area
+                    # if cluster_info.load >= 1:
+                    #     ctx.try_again_later(self.find_cluster_retry_delay)
+                    else:
+                        possible_clusters.add(cluster)
+                except OsbsException:
+                    ctx.try_again_later(self.failure_retry_delay)
+            candidates -= set([c for c in candidates if retry_contexts[c.name].failed])
+            candidates -= possible_clusters
+            # if we have ready clusters, let's use them, if not check the retries
+            if candidates and not possible_clusters:
+                wait_for_any_cluster(retry_contexts) # only if candidate?
```

Comment:
This is fine as it is, but a suggestion for readability is to call this unconditionally at the start of the loop. If any cluster is not in retry-wait it will return immediately.